community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/ConstraintIndexCreator.java

```diff
@@ -273,24 +245,38 @@ private IndexDescriptor getOrCreateUniquenessConstraintIndex( KernelStatement st
             // There's already an index for this schema descriptor, which isn't of the type we're after.
             throw new AlreadyIndexedException( schema, CONSTRAINT_CREATION );
         }
-        return createConstraintIndex( schema );
+        return createConstraintIndexState( schema );
     }
 
-    public IndexDescriptor createConstraintIndex( final LabelSchemaDescriptor schema )
+    IndexConstraintState createConstraintIndexState( final LabelSchemaDescriptor schema ) throws IOException, IndexNotFoundKernelException
     {
-        try ( KernelTransaction transaction =
-                      kernelSupplier.get().newTransaction( KernelTransaction.Type.implicit, AUTH_DISABLED );
-              Statement statement = transaction.acquireStatement() )
+        IndexDescriptor index = IndexDescriptorFactory.uniqueForSchema( schema );
+        long newRuleId = schemaStorage.newRuleId();
+        IndexRule indexRule = IndexRule.indexRule( newRuleId, index,
+                schemaIndexProviderMap.getDefaultProvider().getProviderDescriptor() );
+        indexingService.createIndexes( indexRule );
+        return new IndexConstraintState( indexRule, indexingService.getIndexProxy( indexRule.getId() ) );
```

Comment:
Thinking about concurrency: OK so here we're in a state where the schema label lock is held, we've allocated a ruleId and created an index on disk somewhere, but it's not yet visible in our schema right? Soon after returning from this method the schema label lock will be released, while populating. This will allow for other transactions to come in and create the exact same index that we're creating.

Who will win? Will there be deduplication of multiple equal indexes somewhere?

Also regarding cleanup/crash/failing tx: now there may be index files on disk that aren't tracked by a schema record anywhere. How will that cleanup be managed?